easybuild/easyblocks/p/pytorch.py

```diff
@@ -0,0 +1,176 @@
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.tools.config import build_option
+import easybuild.tools.environment as env
+from easybuild.tools.modules import get_software_root
+from easybuild.tools.systemtools import POWER, get_cpu_architecture
+
+
+class EB_PyTorch(PythonPackage):
+    """Support for building/installing TensorFlow."""
+
+    @staticmethod
+    def extra_options():
+        extra_vars = PythonPackage.extra_options()
+        extra_vars.update({
+            # see https://developer.nvidia.com/cuda-gpus
+            'cuda_compute_capabilities': [[], 'List of CUDA compute capabilities to build with', CUSTOM],
+            'excluded_tests': [{}, 'Mapping of architecture strings to list of tests to be excluded', CUSTOM],
+            'extra_opts': [[], 'List of extra options, overwrites defaults from EasyBlock, e.g. USE_MKLDNN=0', CUSTOM]
+        })
+        extra_vars['download_dep_fail'][0] = True
+        extra_vars['sanity_pip_check'][0] = True
+
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        super(EB_PyTorch, self).__init__(*args, **kwargs)
+        self.options['modulename'] = 'torch'
+        # Test as-if pytorch was installed
+        self.testinstall = True
+        self.tmpdir = tempfile.mkdtemp(suffix='-pytorch-build')
+
+    def fetch_step(self, skip_checksums=False):
+        super(EB_PyTorch, self).fetch_step(skip_checksums)
+        # Resolve tests early to avoid failures later. Use obtain_file if path is not absolute
+        tests = [test if os.path.isabs(test) else self.obtain_file(test) for test in self.cfg['tests']]
+        self.cfg['tests'] = tests
+
+    def configure_step(self):
+        """Custom configure procedure for PyTorch."""
+        super(EB_PyTorch, self).configure_step()
+
+        options = ['PYTORCH_BUILD_VERSION=' + self.version, 'PYTORCH_BUILD_NUMBER=1']
+        # Restrict parallelism
+        options.append('MAX_JOBS=%(parallel)s')
+
+        # BLAS Interface
+        if get_software_root('imkl'):
+            options.append('BLAS=MKL')
+        else:
+            # This is what PyTorch defaults to if no MKL is found. Make this explicit here
+            options.append('BLAS=Eigen')
+
+        known_dependencies = ('FFmpeg', 'gflags', 'glog')
+        dependency_names = set(dep['name'] for dep in self.cfg.dependencies())
+        for dep in known_dependencies:
+            if dep in dependency_names:
+                options.append('USE_%s=1' % dep.upper())
+
+        # Some dependencies that are bundled with PyTorch can be used from the system
+        # No known ECs: 'cpuinfo', 'sleef', 'gloo', 'FP16', 'pthreadpool', 'psimd', 'FXDiv', 'ONNX', 'XNNPack'
+        #               'Eigen', 'GoogleBenchmark'
+        if 'protobuf' in dependency_names:
+            options.append('BUILD_CUSTOM_PROTOBUF=0')
+        # Always use Infiniband
+        options.append('USE_IBVERBS=1')
```

Comment:
Oof, it doesn't seem smart to blindly hardcode that...

Two options here:

* Do it by default, but provide a way out. For example check whether `USE_IBVERBS=` is already specified in `options`, and only add `USE_IBVERBS=1` if it's not. That way, you can opt out by using `extra_opts = ['USE_IBVERBS=0']` in the easyconfig file. That does require moving up the `options.extend(self.cfg['extra_opts'])` to above this part, of course.
  That's probably a good idea to do for all stuff you add to `options`, so introducing a (local) utility function like `add_option('USE_IBVERBS', '1')` probably makes sense?

* Don't do this at all in the easyblock, and leave it up to the easyconfig file to set this via `extra_opts`.

Reply:
Done slightly differently: I'm post-processing the options list into a unique_options list at the end. Makes for more straight-forward code.